magicgui/_qt.py

```diff
@@ -312,3 +336,106 @@ def setValue(self, value):
     def setMaximum(self, value):
         """Set maximum position of slider in float units."""
         super().setMaximum(value * self.PRECISION)
+
+
+class FileDialogMode(Enum):
+    """FileDialog mode options.
+
+    EXISTING_FILE - returns one existing file.
+    EXISTING_FILES - return one or more existing files.
+    OPTIONAL_FILE - return one file name that does not have to exist.
+    EXISTING_DIRECTORY - returns one existing directory.
+    R - read mode, returns one or more existing files.
+        Alias of EXISTING_FILES.
+    W - write mode, returns one file name that does not have to exist.
+        Alias of OPTIONAL_FILE.
+    """
+
+    EXISTING_FILE = "getOpenFileName"
+    EXISTING_FILES = "getOpenFileNames"
+    OPTIONAL_FILE = "getSaveFileName"
+    EXISTING_DIRECTORY = "getExistingDirectory"
+    # Aliases
+    R = "getOpenFileNames"
+    W = "getSaveFileName"
+
+
+class MagicFileDialog(QWidget):
+    """A LineEdit widget with a QFileDialog button."""
+
+    def __init__(
+        self,
+        parent=None,
+        mode: Union[FileDialogMode, str] = FileDialogMode.OPTIONAL_FILE,
+        filter: str = "",
+    ):
+        super().__init__(parent)
+        self.line_edit = QLineEdit(self)
+        self.choose_btn = QPushButton("Choose file", self)
+        self.choose_btn.clicked.connect(self._on_choose_clicked)
+        self.mode = mode
+        self.filter: str = filter
+        layout = QHBoxLayout(self)
+        layout.addWidget(self.line_edit)
+        layout.addWidget(self.choose_btn)
+
+    def _help_text(self):
+        if self.mode is FileDialogMode.EXISTING_DIRECTORY:
+            return "Choose directory"
+        else:
+            return "Select file" + ("s" if self.mode.name.endswith("S") else "")
+
+    @property
+    def mode(self):
+        """Mode for the FileDialog."""
+        return self._mode
+
+    @mode.setter
+    def mode(self, value: Union[FileDialogMode, str]):
+        mode: Union[FileDialogMode, str] = value
+        if isinstance(value, str):
+            try:
+                mode = FileDialogMode(value)
+            except ValueError:
+                try:
+                    mode = FileDialogMode[value.upper()]
+                except KeyError:
+                    pass
```

Comment:
Shouldn't this raise a helpful error message rather than passing silently?

Reply:
I also can't figure out how to trigger this logic branch, in what circumstance does it happen?

Reply:
It will happen if a string is passed that is not a valid choice.  First it checks the enum values, then it checks the enum names.  If neither of those pass, then `mode` will _still_ be a string and then next `ValueError` will be raised (the `pass` let's us write a single `ValueError`).  I added a suggestion for how to test it, by testing an invalid string, not just an invalid type `123`)

Reply:
Thanks!
I think the source of my confusion was assuming the second `if` block was equivalent to an `elif` when I first skimmed over the code. Perhaps if I'd revisited that assumption later I would have caught my own error in logic.

Reply:
yeah, it's also a rather implicitly coded block.  might be nice to add a comment there